internal/internal_workflow.go

```diff
@@ -1095,8 +1104,10 @@ func (d *dispatcherImpl) ExecuteUntilAllBlocked(deadlockDetectionTimeout time.Du
 		d.mutex.Unlock()
 	}()
 	allBlocked := false
+	d.coroutines = append(d.newEagerCoroutines, d.coroutines...)
+	d.newEagerCoroutines = nil
 	// Keep executing until at least one goroutine made some progress
-	for !allBlocked {
+	for !allBlocked || d.allBlockedCallback() {
```

Comment:
nit: this reads a bit oddly ("If not allblocked or allblocked"). I'm trying to think of a better name for the callback. Maybe something like `tryAdvance()`?